services/server/src/server/services/utils/proxy-contract-util.ts

```diff
@@ -0,0 +1,132 @@
+import { whatsabi } from "@shazow/whatsabi";
+import { SourcifyChain } from "@ethereum-sourcify/lib-sourcify";
+import { AbiCoder } from "ethers";
+
+export type ProxyType =
+  | "EIP1167Proxy"
+  | "FixedProxy"
+  | "EIP1967Proxy"
+  | "GnosisSafeProxy"
+  | "DiamondProxy"
+  | "PROXIABLEProxy"
+  | "ZeppelinOSProxy"
+  | "SequenceWalletProxy";
+
+export interface ProxyDetectionResult {
+  isProxy: boolean;
+  proxyType: ProxyType | null;
+  implementations: string[];
+}
+
+export async function detectAndResolveProxy(
+  bytecode: string,
+  address: string,
+  sourcifyChain: SourcifyChain,
+): Promise<ProxyDetectionResult> {
+  // Pass our bytecode to whatsabi so it does not need to fetch it from an rpc
+  const codeCache = {
+    [address]: bytecode,
+  };
+  const cachedCodeProvider = whatsabi.providers.WithCachedCode(
+    // This object mocks a provider as it is not needed for detection only
+    {
+      request: () => {},
+    },
+    codeCache,
+  );
+
+  // Detect proxies but skip other functionalities of whatsabi
+  const detectionResult = await whatsabi.autoload(address, {
+    provider: cachedCodeProvider,
+    abiLoader: false,
+    signatureLookup: false,
+    followProxies: false,
+  });
+
+  // Ignore contracts that have CREATE opcodes as they could falsely be detected as proxies.
+  if (detectionResult.isFactory) {
+    return { isProxy: false, proxyType: null, implementations: [] };
+  }
+
+  const proxyResolvers = detectionResult.proxies;
+
+  // In the following, we check the returned proxy resolvers
+  // and resolve the implementation address for the first valid proxy resolver.
+  // We first handle FixedProxies and DiamondProxies, as their implementations are resolved differently.
+  // Most of the assumptions here are based on the proxy detection experiments in:
+  // https://github.com/sourcifyeth/data-analysis-scripts
+
+  const fixedProxy = proxyResolvers.find(
+    (proxy) => proxy instanceof whatsabi.proxies.FixedProxyResolver,
+  );
+  // Only return EIP1167Proxies because whatsabi can falsely detect non-proxy contracts as FixedProxies (e.g. libraries)
+  if (fixedProxy && isEIP1167Proxy(bytecode, fixedProxy.resolvedAddress)) {
+    return {
+      isProxy: true,
+      proxyType: "EIP1167Proxy",
+      implementations: [fixedProxy.resolvedAddress],
+    };
+  }
```

Comment:
The whatsabi returns a `FixedProxy` for anything called with a `DELEGATECALL`. I think the idea is to also catch non-standardized "minimal" proxies, but the problem with this approach is that it can also include libraries as seen in this issue: https://github.com/shazow/whatsabi/issues/142

Therefore, I decided to only support the standard EIP 1167.